libr/util/log.c

```diff
@@ -31,3 +29,24 @@ R_API void r_log_error(const char *str) {
 R_API void r_log_progress(const char *str, int percent) {
 	printf ("%d%%: %s\n", percent, str);
 }
+
+R_API void r_vlogf(RLogLevel level, const char *fmt, va_list ap) {
+	static const char *headers[R_LOG_MAX_VALUE] = {
+		[R_LOG_DEBUG] = "DEBUG: ",
+		[R_LOG_INFO] = "INFO: ",
+		[R_LOG_WARNING] = "WARNING: ",
+		[R_LOG_ERROR] = "ERROR: ",
+		[R_LOG_CRITICAL] = "CRITICAL: ",
+	};
+
+	const char *hdr = R_BETWEEN (0, level, R_LOG_MAX_VALUE - 1) ? headers[level] : "";
+	eprintf ("%s", hdr);
```

Comment:
why not fputs which makes implicit stderr? anyway, after thinking a bit , even fputs is wrong, because it will be racy and can result in wrong output if multiple threads spit out logs here because this should be one more argument of the format string, not a separate call, but this can be only solved by doing the format string in one place which you refuse

Reply:
As we said in chat, the string in the r_assert_log is a separate "problem". The race condition here does exist and it is about this new log function, which prints both an header of "LogLevel" and a fmt string.

I looked how glib does this and there are two cases (more actually, but these are the one we care about):
1) in cases where the function is called with low memory (infinite recursion and similar corner cases), they just use multiple `printf`
2) in the other cases, they construct a single fmt string on the heap and use just one `printf` to print everything.

I vote for the option 2. We could use the stack, if you prefer, but it seems a good option in both cases to me.

Reply:
Please look at the new commit. I've tried to fix this problem by using a stack-allocated buf which I first fill with the hdr+fmt log and then I print it with only one call to eprintf